Minimal ED64 support

networkfusion opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
Adds the basic needs to run ROMs on V series hardware (specifically the ED64P).

NOTE: this PR will not implement the ability to load the latest firmware and relies upon `libCart` for basic needs.
Improvements will be made in further PR's.


## Motivation and Context
<!--- What does this sample do? What problem does it solve? -->
<!--- If it fixes/closes/resolves an open issue, please link to the issue here  -->
#15 but without all of the fat (needed for the future).

## How Has This Been Tested?
<!-- (if applicable) -->
<!--- Please describe in detail how you tested your sample/changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Screenshots
<!-- (if appropriate): -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Improvement (non-breaking change that adds a new feature)
- [ ] Bug fix (fixes an issue)
- [ ] Breaking change (breaking change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

<!--- It would be nice if you could sign off your contribution by replacing the name with your GitHub user name and GitHub email contact. -->
Signed-off-by: GITHUB_USER <GITHUB_USER_EMAIL>


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **New Features**
	- Added support for the ED64 flashcart, enhancing gameplay for Nintendo 64 games.
	- New instructions for downloading and using ED64 and ED64P ROMs have been documented.
	- Introduced detailed guidelines for uploading ROMs via USB using UNFLoader.

- **Documentation**
	- Expanded the README to clarify installation and usage steps for ED64 and ED64P.
	- New README file specific to the ED64 flashcart added, detailing compatibility and usage.

- **Bug Fixes**
	- Improved initialization logic for different cartridge types to correctly support ED64 functionality.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

networkfusion committed:
Minimal ED64 support

Adds the basic needs to run ROMs on V series hardware.

networkfusion committed:
Merge branch 'main' into ed64-basic

networkfusion committed:
Merge remote-tracking branch 'upstream/main' into ed64-basic

networkfusion committed:
Fix expected save type

networkfusion committed:
Add required changes before merge

networkfusion committed:
Merge remote-tracking branch 'upstream/main' into ed64-basic

ariahirose64 committed:
Fix saves for real

networkfusion committed:
Fix loading.

ariahirose64 committed:
Merge remote-tracking branch 'networkfusion/ed64-basic' into ed64-basic

networkfusion committed:
Further improve ROM loading ability.

Add fixme for why it might be broken.

ariahiro64 committed:
Fix saves on ED64 (#7)

networkfusion committed:
Add fixme notes for save restore

ariahirose64 committed:
Merge remote-tracking branch 'networkfusion/ed64-basic' into HEAD

ariahirose64 committed:
really rough prototype code

networkfusion committed:
Add settings (not used yet).

Improve comments for ED64 saves.

networkfusion committed:
Merge remote-tracking branch 'upstream/main' into ed64-basic

ariahirose64 committed:
bump

ariahirose64 committed:
unless path hardcodded not working

ariahirose64 committed:
fix warnings

ariahirose64 committed:
account for games that dont save

ariahirose64 committed:
working but hardcodded

ariahirose64 committed:
working but needs testing

ariahirose64 committed:
fully working

ariahirose64 committed:
cleaned up code slightly

ariahirose64 committed:
remove types.h

ariahirose64 committed:
remove bloat

ariahirose64 committed:
fix

ariahirose64 committed:
flash ram fixed

ariahirose64 committed:
use libdragon io

ariahiro64 committed:
Update src/menu/cart_load.c

remove whitespace

Co-authored-by: Robin Jones <networkfusion@users.noreply.github.com>

ariahiro64 committed:
Update src/flashcart/ed64/ed64.c

Co-authored-by: Robin Jones <networkfusion@users.noreply.github.com>

ariahiro64 committed:
Update src/flashcart/ed64/ed64.c

Co-authored-by: Robin Jones <networkfusion@users.noreply.github.com>

ariahiro64 committed:
Update src/flashcart/ed64/ed64.c

Co-authored-by: Robin Jones <networkfusion@users.noreply.github.com>

ariahirose64 committed:
fixed sram 128 and added suggested changes

ariahirose64 committed:
dinosaur planet working

ariahirose64 committed:
hotfix

networkfusion committed:
Improve code style

Function names.
file names.
things included in header.

networkfusion committed:
Move sd timings

networkfusion committed:
Remove un-necessary header

networkfusion committed:
Code style

Also fixes ed64_ll_set_ram_bank

networkfusion committed:
Add state ini (not used yet)

prefix temp files with ed_ for the moment.

networkfusion committed:
Move to struct for reset and fram state.

networkfusion committed:
Minor comment improvements

networkfusion committed:
Finish state conversion

for last save path
(hopefully not broken).

networkfusion committed:
Improve error conditions

ariahiro64 committed:
Merge pull request #1 from networkfusion/aria-pr-improvemnets

Improve code style

ariahirose64 committed:
fix large saves

ariahirose64 committed:
calculate fram size without seperate function

ariahirose64 committed:
fix other saves

ariahirose64 committed:
Dinosaur planet 100% working

ariahirose64 committed:
fix other saves 2

ariahirose64 committed:
get rid of unneded function

ariahirose64 committed:
make sure sram128 and fram are 128kb

ariahirose64 committed:
dino fix (cubavore still crashes)

networkfusion committed:
Merge pull request #8 from ariahiro64/ed64-basic

ED64P Saves and virtual writeback

networkfusion committed:
Merge branch 'main' into ed64-basic

networkfusion commented:
@Polprzewodnikowy At this point, please can you provide a review of the changes. I realise it is not perfect, so once a review is performed I propose merging the changes into a branch on your `official` repo for further changes to be based against. Another option is to merge it into main (with some caveats in the readme) and then continue to work on it from there... thoughts?

@ariahiro64 Thanks for the work and testing above my original commit.

networkfusion committed:
Update README.md

ariahirose64 committed:
one last commit

networkfusion committed:
Merge pull request #10 from ariahiro64/ed64-basic

one last commit

networkfusion committed:
Add SAVE_TYPE_DD64_CART_PORT

and Code style improvements

networkfusion committed:
Update libdragon

networkfusion committed:
Merge remote-tracking branch 'upstream/main' into ed64-basic

networkfusion committed:
Update submodules

networkfusion committed:
Revert "Update submodules"

This reverts commit e36b5910c392ee63cc048dba390f85829e8c1695.

networkfusion committed:
Fix ed64_state save in minic

networkfusion committed:
Merge branch 'main' into ed64-basic

networkfusion committed:
Update settings.c

networkfusion committed:
Update settings.c

networkfusion committed:
Update settings.h

networkfusion committed:
Update settings.c

networkfusion committed:
Merge branch 'Polprzewodnikowy:main' into ed64-basic

Polprzewodnikowy commented on src/flashcart/flashcart.c:
```diff
@@ -97,6 +98,7 @@ flashcart_err_t flashcart_init (void) {
 
         case CART_EDX:  // Series X EverDrive-64
```
add `break` here as EDX flashcarts would use ED implementation and that wouldn't work at all

networkfusion replied:
Done.

Polprzewodnikowy commented on src/menu/settings.c:
```diff
@@ -18,6 +18,7 @@ static settings_t init = {
     .bgm_enabled = false,
     .sound_enabled = false,
     .rumble_enabled = true,
+    .auto_firmware_update_enabled = false,
```
All "official" ED64 OS just try to load latest firmware into the cart. Do we really need a switch for that?

Polprzewodnikowy replied:
Also I would strongly advise to not store flashcart specific settings in this file. I would rather use `.ini` file that is created in the `ed64_state.c` to store this switch.

networkfusion replied:
Done.

Polprzewodnikowy commented on src/flashcart/ed64/ed64.c:
```diff
@@ -0,0 +1,313 @@
+        return FLASHCART_ERR_LOAD;
+    }
+
+    size_t save_size = file_get_size(strip_sd_prefix(save_path));
```
No need to add `strip_sd_prefix` here. `file_*` functions in `utils/fs.c` already do that. `strip_sd_prefix` should be used only where function expects path to not include `sd:/` prefix, and that generally applies only to the FatFs functions.

Polprzewodnikowy replied:
By the way, if you opened file with `f_open` you already have file size accessible with `f_size(&fil);`. See SC64 implementation: https://github.com/Polprzewodnikowy/N64FlashcartMenu/blob/553e986a380b9ed1cb81253c644ac42ed3ec17b2/src/flashcart/sc64/sc64.c#L418

Polprzewodnikowy commented on src/flashcart/ed64/ed64.c:
```diff
@@ -0,0 +1,313 @@
+    if (f_close(&fil) != FR_OK) {
+        return FLASHCART_ERR_LOAD;
+    }
+
```
You should check if `br` equals to `save_size`, see: https://github.com/Polprzewodnikowy/N64FlashcartMenu/blob/553e986a380b9ed1cb81253c644ac42ed3ec17b2/src/flashcart/sc64/sc64.c#L429-L431

Polprzewodnikowy commented on src/flashcart/ed64/ed64.c:
```diff
@@ -0,0 +1,313 @@
+    case FLASHCART_SAVE_TYPE_SRAM:
+        type = SAVE_TYPE_SRAM;
+        break;
+    case FLASHCART_SAVE_TYPE_SRAM_BANKED:
```
We should add "SRAM banked" support to the `has_feature` function as this is very wrong and won't work with this flashcart. ED64 registers occupy last bank of this save type and that could lead to some undesired behavior. I don't really know right now if we should block loading such save types or just show a warning that this save isn't supported and let user load ROM regardless.

networkfusion replied:
I need to check more fully, but the carts would support it, since a) this menu does not write to it. b) if it did, we should store in the pseudo region for flashram.

Polprzewodnikowy commented on src/flashcart/ed64/ed64_ll.c:
```diff
@@ -0,0 +1,409 @@
+}
+
+
+void ed64_ll_set_sdcard_timing (void) {
```
libcart already does this, not needed

Polprzewodnikowy commented on src/flashcart/ed64/ed64_ll.c:
```diff
@@ -0,0 +1,409 @@
+
+void ed64_ll_reg_write (uint32_t reg, uint32_t data) {
+
+    *(volatile uint32_t *) (ED64_CONFIG_REGS_BASE);
```
Why is this needed, this dummy read is probably optimized away by the compiler anyway

Polprzewodnikowy commented on src/flashcart/ed64/ed64_ll.c:
```diff
@@ -0,0 +1,409 @@
+
+    *(volatile uint32_t *) (ED64_CONFIG_REGS_BASE);
+    *(volatile uint32_t *) (ED64_CONFIG_REGS_BASE + reg * 4) = data;
+    *(volatile uint32_t *) (ROM_ADDRESS);
```
Same as 2 lines above, probably dead code

Polprzewodnikowy commented on src/flashcart/ed64/ed64_ll.c:
```diff
@@ -0,0 +1,409 @@
+void ed64_ll_reg_write (uint32_t reg, uint32_t data) {
+
+    *(volatile uint32_t *) (ED64_CONFIG_REGS_BASE);
+    *(volatile uint32_t *) (ED64_CONFIG_REGS_BASE + reg * 4) = data;
```
Please use `io_write`/`io_read` functions provided by the libdragon

Polprzewodnikowy commented on src/flashcart/ed64/ed64_ll.c:
```diff
@@ -0,0 +1,409 @@
+
+uint32_t ed64_ll_reg_read (uint32_t reg) {
+
+    *(volatile uint32_t *) (ED64_CONFIG_REGS_BASE);
```
Same as in line 63

Polprzewodnikowy commented on src/flashcart/ed64/ed64_ll.c:
```diff
@@ -0,0 +1,409 @@
+uint32_t ed64_ll_reg_read (uint32_t reg) {
+
+    *(volatile uint32_t *) (ED64_CONFIG_REGS_BASE);
+    return *(volatile uint32_t *) (ED64_CONFIG_REGS_BASE + reg * 4);
```
Same as in line 64

Polprzewodnikowy commented on src/flashcart/ed64/ed64_ll.c:
```diff
@@ -0,0 +1,409 @@
+}
+
+
+void pi_initialize (void) {
```
this is completely unnecessary

Polprzewodnikowy commented on src/flashcart/ed64/ed64_ll.c:
```diff
@@ -0,0 +1,409 @@
+}
+
+// Inits PI for sram transfer
+void pi_initialize_sram (void) {
```
these timings should be already initialized by the libcart

Polprzewodnikowy commented on src/flashcart/ed64/ed64_ll.c:
```diff
@@ -0,0 +1,409 @@
+
+}
+
+void pi_dma_from_sram (void *dest, unsigned long offset, unsigned long size) {
```
please use `pi_dma_read_data`/`pi_dma_write_data` functions from `flashcart_utils.c` instead of these functions that do not use libdragon provided implementations

ariahiro64 replied:
the reason i didnt use the libdragon implimentation is because the sram address is write protected (i think). it results in this 
![image](https://github.com/Polprzewodnikowy/N64FlashcartMenu/assets/45049787/a7632fec-a610-4cda-919e-0ba1aec9a2db)

networkfusion replied:
After adding `#include "../flashcart_utils.h"`
Try using `pi_dma_read_data((void*)(ED64_SAVE_ADDR_BASE + address_offset), buffer, size);`

Polprzewodnikowy commented on src/flashcart/ed64/ed64_ll.c:
```diff
@@ -0,0 +1,409 @@
+}
+
+
+void pi_dma_to_sram (void *src, unsigned long offset, unsigned long size) {
```
same as above function

Polprzewodnikowy commented on src/flashcart/ed64/ed64_ll.c:
```diff
@@ -0,0 +1,409 @@
+
+}
+
+void pi_dma_from_cart (void* dest, void* src, unsigned long size) {
```
same as above

Polprzewodnikowy commented on src/flashcart/ed64/ed64_ll.c:
```diff
@@ -0,0 +1,409 @@
+}
+
+
+void pi_dma_to_cart (void* dest, void* src, unsigned long size) {
```
same as above

Polprzewodnikowy commented on src/flashcart/ed64/ed64_ll.c:
```diff
@@ -0,0 +1,409 @@
+
+
+// Wrapper to support unaligned access to memory
+void pi_dma_from_cart_safe (void *dest, void *src, unsigned long size) {
```
same as above, `pi_dma_read_data` already supports unaligned access

Polprzewodnikowy commented on src/flashcart/ed64/ed64_ll.c:
```diff
@@ -0,0 +1,409 @@
+
+    dma_wait();
+
+    io_write(PI_BSD_DOM2_LAT_REG, 0x05);
```
don't really know why it's necessary to reconfigure PI timings, libcart already set these to correct values

Polprzewodnikowy commented on src/flashcart/ed64/ed64_ll.c:
```diff
@@ -0,0 +1,409 @@
+
+    dma_wait();
+
+    pi_initialize();
```
not needed

Polprzewodnikowy commented on src/flashcart/ed64/ed64_ll.c:
```diff
@@ -0,0 +1,409 @@
+
+    dma_wait();
+
+    io_write(PI_BSD_DOM2_LAT_REG, 0x40);
```
same as above, no need  to reconfigure timings

Polprzewodnikowy commented on src/flashcart/ed64/ed64_ll.c:
```diff
@@ -0,0 +1,409 @@
+
+    // Offset Large RAM size 128KiB with a 16KiB nudge to allocate enough space
+    // We do this because 128Kib is not recognised and a 32KiB allocates too little 
+    pi_dma_from_sram(buffer, -(size - KiB(16)), size);
```
What is happening here? Why second argument is negative? I really don't understand what this offset does.

Polprzewodnikowy commented on src/flashcart/ed64/ed64_ll.c:
```diff
@@ -0,0 +1,409 @@
+
+    dma_wait();
+
+    io_write(PI_BSD_DOM2_LAT_REG, 0x05);
```
same as in `ed64_ll_get_sram_128`

Polprzewodnikowy commented on src/flashcart/ed64/ed64_ll.c:
```diff
@@ -0,0 +1,409 @@
+
+    dma_wait();
+
+    io_write(PI_BSD_DOM2_LAT_REG, 0x40);
```
same as in `ed64_ll_get_sram_128`

Polprzewodnikowy commented on src/flashcart/ed64/ed64_ll.c:
```diff
@@ -0,0 +1,409 @@
+    dma_wait();
+    ed64_ll_set_save_type(SAVE_TYPE_FLASHRAM);
+
+    return 1;
```
Just general note to most functions here, why return anything when these functions can't fail?

Polprzewodnikowy commented on src/flashcart/ed64/ed64_ll.c:
```diff
@@ -0,0 +1,409 @@
+    //half working
+    dma_wait();
+    //Timing
+    pi_initialize_sram();
```
Are DOM2 timings set by the libcart too fast to handle read/write to SRAM region? If not then it's not necessary to reconfigure timings.

networkfusion replied:
it does seem they are. the 2.12 Vseries source reverts and then re-applys these timings.

Polprzewodnikowy commented on src/flashcart/ed64/ed64_ll.c:
```diff
@@ -0,0 +1,409 @@
+    dma_wait();
+
+    ed64_ll_set_sram_128(buffer, size);
+    data_cache_hit_writeback_invalidate(buffer, size);
```
Cache writeback is in wrong place, it should happen before doing write in the `ed64_ll_set_sram_128` function. This applies to all functions above.

ariahiro64 commented:
alot of this is me basing it off of legacy code. i also have flash ram and re2 working now.

networkfusion committed:
Add save fixes

Changes from https://github.com/ariahiro64/N64FlashcartMenu/commit/96efc3824e20096dd7f1e28b0e48b1539eb5de65

networkfusion committed:
Minor improvements

Remove changes to settings
Only target Vseries and ED64P

Polprzewodnikowy commented:
> alot of this is me basing it off of legacy code. i also have flash ram and re2 working now.

Nothing wrong with using already existing code, but while we at it, now it's a good moment to make code even better. We know much more about N64 internals and libdragon/libcart is already lifting much work required by menu code.

ariahiro64 commented:
im more wondering what the unicorn 1Mbit SRAM is and is it just a flashcart thing?

networkfusion commented:
> im more wondering what the unicorn 1Mbit SRAM is and is it just a flashcart thing?

Yes, it is a `krikzz` thing. since SRAM is easier than `FRAM` and krikzz decided to do his own thing (as usual).

networkfusion commented:
> > alot of this is me basing it off of legacy code. i also have flash ram and re2 working now.
> 
> Nothing wrong with using already existing code, but while we at it, now it's a good moment to make code even better. We know much more about N64 internals and libdragon/libcart is already lifting much work required by menu code.

I have attempted to make improvements in https://github.com/networkfusion/N64FlashcartMenu/tree/ed64p-refactor but feel we should merge a working version first.

networkfusion commented on src/flashcart/ed64/ed64_ll.c:
```diff
@@ -0,0 +1,326 @@
+//FIXME find a better solution, perferably a libdragon one
+void sleep (unsigned long ms);
+
+void sleep (unsigned long ms) {
```
you can use ` wait_ms()`

networkfusion committed:
Merge branch 'main' into ed64-basic

SlaughterRidge commented:
As per @networkfusion's request, I am commenting on this PR instead of the issue. 

Tested this on EverDrive 2.0 running firmware 2.33

Loading roms appear to work. Saving has been tested and works with EEPROM 4K (Goldeneye), EEPROM 16K (Starwars Episode 1 Racer). Saving doesn't appear to work with SRAM or FRAM. (Ocarina of Time, Harvest Moon, Paper Mario, Majora's Mask) 

I do not own a controller pak to test automatic save backup and loading. (assuming this is a feature of the N64 Flashcart Menu)

Saving on the EverDrive 2.0 requires pushing the reset button on the console to trigger the everdrive to write the save file to the SD card.

networkfusion committed:
Merge branch 'main' into ed64-basic

networkfusion committed:
Merge branch 'main' into ed64-basic

networkfusion committed:
Use libdragon function for sleep

networkfusion committed:
Merge branch 'main' into ed64-basic

networkfusion committed:
Improve readme

Specifically for ED64 V3

networkfusion committed:
Ignore whole tools dir

networkfusion committed:
fix ability to load ROMs

Workaround for need of F/W 3.06 and UNFLoader.
Fixes ability to run ROM's

networkfusion committed:
Merge branch 'main' into ed64-basic

networkfusion committed:
Update ed64.c

change function name to match main branch.

networkfusion committed:
Update ed64.c

fix further function that was renamed.

evilgenius31 commented:
> https://github.com/networkfusion/N64FlashcartMenu/tree/ed64p-refactor

How far should loading roms get with new builds for the ed64p? Menus look great

networkfusion committed:
Update ed64.c

Add all current features to switch.

networkfusion commented:
> > https://github.com/networkfusion/N64FlashcartMenu/tree/ed64p-refactor
> 
> How far should loading roms get with new builds for the ed64p? Menus look great

At minimum, loads clays (oss rom) and blues brothers 2000. Saves are probably broken now. Just tested on my ED64 V3.

networkfusion committed:
Fix hidden ED dirs

networkfusion committed:
Improve state file path

networkfusion committed:
Minor comment improvements and fixes

networkfusion committed:
Merge remote-tracking branch 'upstream/main' into ed64-basic

networkfusion committed:
Add commented out code for ed64x

networkfusion committed:
Merge remote-tracking branch 'upstream/main' into ed64-basic

Matsilagi commented:
As of right now, the writebacks for ED64P are completely broken.
It writes the initial one, but sometimes fails even to return to the main directory. It also will only read (not write) existing ones and not touch any changes.
The last working version from my tests was dated December 24th 2023.

networkfusion committed:
Merge branch 'develop' into ed64-basic

networkfusion committed:
Merge branch 'develop' into ed64-basic

networkfusion committed:
Update ed64.c

Save corrections

networkfusion commented:
@coderabbitai review

networkfusion committed:
Update ed64_ll.h

Improve commenting.

networkfusion committed:
Merge branch 'develop' into ed64-basic

networkfusion committed:
Merge branch 'develop' into ed64-basic

networkfusion committed:
Merge branch 'develop' into ed64-basic

networkfusion committed:
Update README.md

networkfusion committed:
Update README.md

networkfusion committed:
Merge branch 'develop' into ed64-basic

networkfusion committed:
Merge branch 'develop' into ed64-basic
